### PR TITLE
Fix `useResizeObserver` bugs

### PR DESCRIPTION
**Pull request by DaniGuardiola**
## What?

- Fixed a circular dependency issue ([original comment](https://github.com/WordPress/gutenberg/pull/64943#discussion_r1751417597))
- Fixed a bug that caused elements to not be properly unobserved ([original comment](https://github.com/WordPress/gutenberg/pull/64943#discussion_r1751414803))
- Renamed the `_legacy` dir to just `legacy`

Thanks @jsnajdr! Original post-merge review: https://github.com/WordPress/gutenberg/pull/64943#pullrequestreview-2291597248

### CODE REVIEW

**Comment by DaniGuardiola** [comment]
Following up on the discussion here: https://github.com/WordPress/gutenberg/pull/64943#pullrequestreview-2291597248

cc @jsnajdr 

Regarding React Native, I don't think this utility should try to cover that use case. `ResizeObserver` is not a thing in React Native afaik. At most, we can create an API that's somewhat similar to the legacy one (e.g. `useTrackElementSize` or something like that) so it can be used from React Native, but the new one should just be web-only since that's the only context where `ResizeObserver` exists. FWIW, my assumption is that the legacy API will still work in React Native so nothing should break.

File: packages/compose/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+### Bug Fixes
+
+-   `useResizeObserver`: fixed a bug that caused elements to not be properly unobserved ([#65389](https://github.com/WordPress/gutenberg/pull/65389)).
+
```
Comment (jsnajdr) [inline packages/compose/CHANGELOG.md]: If the original `useResizeObserver` is not yet released, then it doesn't need a changelog entry. The package user never had a chance to observe the bugs.

File: packages/compose/src/hooks/use-resize-observer/use-resize-observer.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+		}
+
+		// Set up `ResizeObserver`.
+		resizeObserverRef.current ??= new ResizeObserver( callbackEvent );
+		const { current: resizeObserver } = resizeObserverRef;
+
+		// Unobserve previous element.
+		if ( observedElementRef.current ) {
+			resizeObserver.unobserve( observedElementRef.current );
+		}
```
Comment (jsnajdr) [inline packages/compose/src/hooks/use-resize-observer/use-resize-observer.ts:35]: Nice fix, this should indeed unobserve the previous element properly.

**Review by jsnajdr — APPROVED** [review]

**Comment by jsnajdr** [comment]
> Regarding React Native, I don't think this utility should try to cover that use case. `ResizeObserver` is not a thing in React Native afaik.

What if React Native module imports
```js
import { useResizeObserver } from '@wordpress/compose';
```
? Then it also imports the web version of the new hook, and that hook can be called as `useResizeObserver( callback )`, and it will execute the web version and will probably fail. That's what I'm worried about. Not that we don't implement a RN version of the new hook, but that we're incorrectly exposing the new version of it. The web and RN versions should be better isolated from each other.

File: packages/compose/CHANGELOG.md (hunk 1 above)
Comment (DaniGuardiola) [reply packages/compose/CHANGELOG.md]: True, I thought about this but decided to add the entry anyway, however you're right so I'll remove it!

**Comment by DaniGuardiola** [comment]
I see the point about React Native, however I wonder, do all `@wordpress/compose` utilities guarantee that they will function in a React Native context? I would be surprised if so. In this case, I guess there are two considerations:

- Since the previous API did support React Native, should we publish a dev note or similar to give the community a heads up?
- Should we implement an API that does work in React Native too, like I was proposing? Is this a recurrent enough need? Is there an advantage in having a "homomorphic" utility that works in both web and React Native with the same API?

What do you think?

**Comment by jsnajdr** [comment]
> do all `@wordpress/compose` utilities guarantee that they will function in a React Native context?

There are two separate index files, `index.js` and `index.native.js`, and each exports only the utils that are relevant for a given platform. So, yes, there is such a guarantee.

All I'm asking for is that there is a new `use-resize-observer/index.native.js` file that exports only the `legacy/index.native.js` hook. Don't export the web version at all.

You don't need to write any new code or remove anything. Just exclude the web version of `useResizeObserver( callback )` from the RN version. The RN version should be completely untouched by this addition.

**Comment by tyxla** [comment]
> > do all `@wordpress/compose` utilities guarantee that they will function in a React Native context?
> 
> There are two separate index files, `index.js` and `index.native.js`, and each exports only the utils that are relevant for a given platform. So, yes, there is such a guarantee.
> 
> All I'm asking for is that there is a new `use-resize-observer/index.native.js` file that exports only the `legacy/index.native.js` hook. Don't export the web version at all.
> 
> You don't need to write any new code or remove anything. Just exclude the web version of `useResizeObserver( callback )` from the RN version. The RN version should be completely untouched by this addition.

This makes sense to me, but if in doubt, we can always ping some folks from the mobile team for confirmation.

File: packages/compose/CHANGELOG.md (hunk 1 above)
Comment (DaniGuardiola) [reply packages/compose/CHANGELOG.md]: Removed

**Comment by DaniGuardiola** [comment]
@jsnajdr alright, that makes sense though there are some details to discuss. For example, what happens with the difference in API. The new API is web-only so it won't be possible to re-create for React Native in any case. Will that be okay?

Anyway, I'm merging this but let's keep discussing and I'll open a new PR to handle that.

**Comment by jsnajdr** [comment]
> For example, what happens with the difference in API. The new API is web-only so it won't be possible to re-create for React Native in any case. Will that be okay?

At this moment, the React Native API stays the same. There is the `useResizeObserver` hook that returns a helper element. I'm only proposing that we stop breaking that API by exposing a web version of the new API that doesn't work in RN anyway.

It should be possible to implement a pretty good version of the new API also for React Native. `View` elements in RN have an `onLayout` event handler, and we can use it, something like this:
```js
function useResizeObserver( callback ) {
  const elRef = useRef();
  const cb = useEvent( callback );
  useEffect( () => {
    function handleLayout( e ) {
      const entries = [ { contentBoxSize: [ { inlineSize: e.layout.width, blockSize: e.layout.height } ] } ];
      cb( entries );
    }
    const el = elRef.current;
    el.addEventListener( 'layout', handleLayout );
    return () => el.removeEventListener( 'layout', handleLayout );
  } );
  return elRef;
}
```
The existing RN usages of `useResizeObserver` that I've found should be happy with this.

There are details to figure out like whether the reported layout is `contentBoxSize` or rather `borderBoxSize`. I don't know how that works in RN, or even if there is such a thing as borders and paddings and margins there at all.

**Comment by ciampo** [comment]
Just to understand, with this PR merged, is there a chance that the native APIs got a breaking change? If so, we should make sure we quickly put up a PR to export only the legacy version to react native, as proposed by Jarda above.

**Comment by DaniGuardiola** [comment]
@ciampo nothing that already is in use will break. I will put up a PR that handles the React Native situation very soon, but don't worry about it as there's no chance new code will be submitted that breaks, as the error would be immediate.

This PR doesn't affect the React Native situation fwiw, it just addresses some bugs and tweaks related to the web version.

**Comment by DaniGuardiola** [comment]
See https://github.com/WordPress/gutenberg/pull/65588

**Pull request merged**